java/com/google/re2j/RE2.java

```diff
@@ -110,8 +110,13 @@
   int prefixRune;               // first rune in prefix
 
   // Cache of machines for running regexp.
-  // Accesses must be serialized using |this| monitor.
-  private final List<Machine> machine = new ArrayList<Machine>();
+  private final ThreadLocal<Machine> machineThreadLocal =
```

Comment:
I wonder: would sharing this cache across all instances of RE2 reduce allocation still further?  Each regexp has its own per-thread cache of machines, but that means, looking across a set of compiled regexps, all but one machine for a given thread are necessarily sitting idle.

You'd need to make this field static, and change Machine slightly so that it could be reused for different RE2 instances.

Your existing benchmarks would probably not reveal a change, but any program in which a thread executes one regexp followed by a different one would benefit.

Reply:
The gain is minimal since you'll need to create two new instances of Queue per RE2. 

Also add an additional check  like:

```
init (re2){
 if this.re2 != re2 {
      this.q0 = new Queue...
      this.q1 = new Queue...
      this.re2=re2;
  }
}
```

Reply:
Right, the queue is tailored to the machine.  Never mind.